opal/corelib/array.rb

```diff
@@ -1136,40 +1136,39 @@ def flatten!(level = undefined)
 
   def hash
     %x{
-      var top = (Opal.hash_ids == undefined),
+      var top = (Opal.hash_ids === undefined),
           result = ['A'],
-          hash_id = self.$object_id(),
-          item, i, key;
+          __id__ = self.$$id || self.$__id__(),
+          item, i, key, self_key = 's';
 
       try {
         if (top) {
-          Opal.hash_ids = {};
+          Opal.hash_ids = Object.create(null);
         }
 
-        if (Opal.hash_ids.hasOwnProperty(hash_id)) {
-          return 'self';
+        // return early for recursive structures
+        if (Opal.hash_ids[__id__]) {
+          return self_key;
         }
 
         for (key in Opal.hash_ids) {
-          if (Opal.hash_ids.hasOwnProperty(key)) {
-            item = Opal.hash_ids[key];
-            if (#{eql?(`item`)}) {
-              return 'self';
-            }
+          item = Opal.hash_ids[key];
+          if (#{eql?(`item`)}) {
+            return self_key;
           }
         }
 
-        Opal.hash_ids[hash_id] = self;
+        Opal.hash_ids[__id__] = true;
 
         for (i = 0; i < self.length; i++) {
           item = self[i];
-          result.push(item.$hash());
+          result.push(item.$hash(), ',');
         }
 
-        return result.join(',');
+        return result.join();
       } finally {
         if (top) {
-          delete Opal.hash_ids;
+          Opal.hash_ids = undefined;
```

Comment:
Deleting a property causes the VM to put it in "dictionary" or "hash table" mode. Iterating over it with `for(var foo in bar)` triggers a de-opt for that function. :-\ You'd think that, based on what those words mean, objects created with `Object.create(null)` would be in this mode by default, but I guess they aren't until you `delete` from them.

https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#52-the-object-being-iterated-is-not-a-simple-enumerable